packages/SwingSet/test/vat-admin/terminate/test-terminate.js

```diff
@@ -34,9 +43,77 @@ test('terminate', async t => {
     'GOT QUERY 3',
     'foreverP.catch vat terminated',
     'query3P.catch vat terminated',
-    'foo4P.catch vat is dead',
+    'foo4P.catch unknown vat',
```

Comment:
Oh, wait, if other vats can distinguish the difference between a vat being recently dead (`ephemeral.vats.get(vatID).dead`) and a vat being missing because we've reloaded the kernel since then (`!ephemeral.has(vatID)`), then that's a consensus failure waiting to happen. One validator might reboot before the message is sent, the other might not, and they get different messages.

I can't immediately tell whether this change indicates that things might be ambiguous.. please doublecheck.

Reply:
Yes, it was the difference in behavior between first run and reply that lead me to change the error message, so that it wouldn't allow an observer to distinguish the two cases.  Also, I don't think it's a message sender's business whether a vat is dead or just never existed in the first place.

Reply:
Good, thanks.

I fully agree that callers must not be able to tell the difference between "that vat died recently" and "that vat died and then we rebooted", because orthogonal persistence means that vats must not be able to sense reboots.

But if the vat *really* never existed in the first place, not even in a previous reboot, then nobody should have ever been able to send anything to it, and the vat shouldn't even have a way to ask the question.

Reply:
If the vat is not running liveslots it can trivially generate a reference to a vat that doesn't exist.  Arguably that should be treated as misbehavior that kills the vat (actually, before the most recent round of changes, such an act would instead panic the kernel, which is almost certainly the wrong thing).  My original motivation for not making such a distinction on the kernel side was that I was concerned about a monotonically increasing memory burden of keeping track of which vats were dead, but it just occurred to me that if we assign vat IDs from a monotonically increasing counter (which we do), then any unknown vat ID that is less that the value of the current vat ID allocation count is a dead vat and any vat ID greater than the count is a non-existent vat, so I now actually think we could implement this trivially.  However, this now opens up a path for a mis-behaved vat to inappropriately probe the kernel as to the state of the kernel's world.  If we kill a vat for referring to a non-existent (as opposed to dead) vat ID, then this is only an anthropic channel so I'm not sure if it's an issue or not, but I don't want to just assume that it wouldn't be a problem without deeper analysis (for example: a worker that after doing all its regular work, as a last act tries to send a message to vat N, where N is some threshold value, and if it's still alive afterward sends a message to a known outsider that sets off a bomb destroying the data center).

Reply:
On further reflection, I think some of what I just said may be wrong.  When a vat talks to a kernel, it never makes direct reference to a vat ID outside itself, does it?  That is always mediated via a ctable lookup.  The issue is that we found there are ways to hold onto a legitimate entity in kernel space (like an unresolved promise) that itself encapsulates a reference to a vat ID that is no longer valid, but you can't ask about vat IDs directly.  Is that right?

Reply:
yep, exactly

Once the c-list row is removed, the sending vat no longer has a way to reference the old kernel object/promise id, no matter how hard they try. They might reference some vat-side vpid/void which isn't in the c-list, but that's their fault, and vat-fatal. So even if the kernel object table maps e.g. `ko12` to `v4 (dead)`, then once we've removed all the c-list entries that reference `ko12`, there's nothing a vat can do to create new references to `ko12`. But, as you pointed out, there may still be `v4` references lying around *in the kernel*, and until we know that we've cleaned out all of those (by deleting entries, or replacing e.g. `ko12 -> v4` with `ko12 -> null`), we need to be able to survive them getting used somehow.

If we look at just the nominal type signatures of our kernel tables, I think the only places where vatids show up are:

* kernel object table, `koNN -> vNN`
* decider property of unresolved promises in the kernel promise table
* subscriber list of unresolved promises
* target vatid of notification items in the run-queue

I think we've already removed the decider property, by virtue of rejecting all the dead vat's promises, thus converting those unresolved promises into resolved ones (which means they no longer have a `.decider` property).

We currently tolerate the presence of dead vats in in the notification run-queue items by ignoring the delivery when it reaches the top of the run-queue. This also lets us tolerate the dead vat having subscribed to promises, which will cause the same kind of notifications. If we wanted to be preemptive, we could walk all unresolved promises and filter their subscriber lists (to remove the dead vat), and walk the run-queue and filter the notifications, and then we wouldn't have to tolerate those stale notifications reaching the front of the run-queue because we'd already deleted them all.

We tolerate the dead vats in the kernel object table by rejecting messages during `deliverToVat` when the target is an object which that table maps to a dead vat. We must retain that rejection step (the dull thud of messages bouncing off the closed door), but we could trigger it by the OT lookup returning `null` instead of a vatID which is then looked up in a separate alive/dead table. To do that, we'd have to walk the the dead vat's c-list for all the objects it had exported into the kernel (`o+NN`, not `o-NN`), find the `koNN` they map to, and then edit those in the kernel object table. This should be equivalent to walking the entire kernel object table for all the entries that map to `vNN` and editing them (and a lot cheaper, at least until we have some sort of index on the OT), but it might be nice to have some sort of invariant check as well.

It's tempting to consider a step after that, to walk all the *other* c-list and find entries that point to dead objects, and replace their targets with `null` (e.g. vat-alice would have a c-list that starts with `o-4 -> ko12`, but ko12 was in the dead vat, so vat-alice is left with a c-list of `o-4 -> null`), and delete the OT entry entirely. But I think we can't do that because objects, even dead ones, retain identity, and vats need a way to talk *about* (but not *to*) the dead. Which I think means we're stuck with retaining the `ko12` in the OT as long as any vat holds a reference to it.

So we can purge `vNN` from the kernel entirely, but not the object IDs that used to live in it.